File: Python/Interview_problems/File_System_demo.py
"""
================================================================================
LLD: In-Memory File System
================================================================================

Functional Requirements:
    1. Create a directory at a given path.
    2. Create a file at a given path.
    3. Write (overwrite) content to a file.
    4. Read the content of a file.
    5. Delete a file or a directory (recursively).
    6. Calculate the size of a file or a directory (recursive sum).
    7. Move a file or directory from one path to another.
    8. Enforce permissions at the directory level for ADMIN vs USER.

Non-Functional Requirements:
    1. Extensibility: adding a new node type (e.g. symlink) shouldn't
       require touching path-resolution logic.
    2. Predictable errors: an operation on a path that doesn't exist, or
       one a user lacks permission for, fails with a clear exception
       instead of corrupting the tree.

Design:
    FileSystemNode (abstract) is the common base for File and Directory --
    both know their name, parent, and how to report their own size.
    Directory additionally holds a name -> FileSystemNode map of children
    and an optional `required_role`: an operation that mutates something
    under this directory (or the directory itself) must be performed by a
    user whose role is at least `required_role`.

    FileSystemManager resolves "/"-separated paths by walking the tree
    from the root and performs permission checks against the *parent*
    directory of the node being mutated (create/write/delete/move all
    change the parent's contents).

Core Entities:
    UserType / User
    FileSystemNode (abstract) / File / Directory
    FileSystemManager
================================================================================
"""

from abc import ABC, abstractmethod
from enum import IntEnum
from typing import Dict, Optional


class UserType(IntEnum):
    USER = 0
    ADMIN = 1


class User:
    def __init__(self, name: str, role: UserType):
        self.name = name
        self.role = role


class PathNotFoundError(RuntimeError):
    pass


class PathAlreadyExistsError(RuntimeError):
    pass


class PermissionDeniedError(RuntimeError):
    pass


class FileSystemNode(ABC):
    def __init__(self, name: str, parent: Optional["Directory"]):
        self.name = name
        self.parent = parent

    @abstractmethod
    def get_size(self) -> int:
        raise NotImplementedError

    @abstractmethod
    def is_directory(self) -> bool:
        raise NotImplementedError


class File(FileSystemNode):
    def __init__(self, name: str, parent: Optional["Directory"]):
        super().__init__(name, parent)
        self.content = ""

    def read(self) -> str:
        return self.content

    def write(self, content: str) -> None:
        self.content = content

    def append(self, content: str) -> None:
        self.content += content

    def get_size(self) -> int:
        return len(self.content)

    def is_directory(self) -> bool:
        return False


class Directory(FileSystemNode):
    def __init__(self, name: str, parent: Optional["Directory"], required_role: UserType = UserType.USER):
        super().__init__(name, parent)
        self.children: Dict[str, FileSystemNode] = {}
        self.required_role = required_role

    def add_child(self, child: FileSystemNode) -> None:
        self.children[child.name] = child

    def get_child(self, name: str) -> Optional[FileSystemNode]:
        return self.children.get(name)

    def remove_child(self, name: str) -> Optional[FileSystemNode]:
        return self.children.pop(name, None)

    def get_size(self) -> int:
        return sum(child.get_size() for child in self.children.values())

    def is_directory(self) -> bool:
        return True


def _split_path(path: str):
    parts = [part for part in path.split("/") if part]
    if not parts:
        raise ValueError("Path must not be empty")
    return parts


class FileSystemManager:
    def __init__(self):
        self.root = Directory("/", None)

    def _resolve_parent(self, parts) -> Directory:
        current = self.root
        for part in parts[:-1]:
            child = current.get_child(part)
            if child is None or not child.is_directory():
                raise PathNotFoundError(f"No such directory: {part}")
            current = child
        return current

    def _resolve(self, path: str) -> FileSystemNode:
        parts = _split_path(path)
        current: FileSystemNode = self.root
        for part in parts:
            if not current.is_directory():
                raise PathNotFoundError(f"Not a directory: {current.name}")
            child = current.get_child(part)
            if child is None:
                raise PathNotFoundError(f"No such path: {path}")
            current = child
        return current

    def _check_permission(self, directory: Directory, user: User) -> None:
        if user.role < directory.required_role:
            raise PermissionDeniedError(
                f"User {user.name} ({user.role.name}) lacks permission for {directory.name} "
                f"(requires {directory.required_role.name})")

    def create_directory(self, user: User, path: str, required_role: UserType = UserType.USER) -> Directory:
        parts = _split_path(path)
        parent = self._resolve_parent(parts)
        self._check_permission(parent, user)
        name = parts[-1]
        if parent.get_child(name) is not None:
            raise PathAlreadyExistsError(f"Path already exists: {path}")
        directory = Directory(name, parent, required_role)
        parent.add_child(directory)
        return directory

    def create_file(self, user: User, path: str) -> File:
        parts = _split_path(path)
        parent = self._resolve_parent(parts)
        self._check_permission(parent, user)
        name = parts[-1]
        if parent.get_child(name) is not None:
            raise PathAlreadyExistsError(f"Path already exists: {path}")
        file = File(name, parent)
        parent.add_child(file)
        return file

    def write_file(self, user: User, path: str, content: str) -> None:
        node = self._resolve(path)
        if not isinstance(node, File):
            raise ValueError(f"Not a file: {path}")
        self._check_permission(node.parent, user)
        node.write(content)

    def read_file(self, user: User, path: str) -> str:
        node = self._resolve(path)
        if not isinstance(node, File):
            raise ValueError(f"Not a file: {path}")
        return node.read()

    def delete(self, user: User, path: str) -> None:
        node = self._resolve(path)
        parent = node.parent
        if parent is None:
            raise ValueError("Cannot delete the root directory")
        self._check_permission(parent, user)
        parent.remove_child(node.name)

    def get_size(self, path: str) -> int:
        return self._resolve(path).get_size()

    def move(self, user: User, source: str, destination: str) -> None:
        node = self._resolve(source)
        old_parent = node.parent
        if old_parent is None:
            raise ValueError("Cannot move the root directory")
        self._check_permission(old_parent, user)

        dest_parts = _split_path(destination)
        new_parent = self._resolve_parent(dest_parts)
        self._check_permission(new_parent, user)
        new_name = dest_parts[-1]
        if new_parent.get_child(new_name) is not None:
            raise PathAlreadyExistsError(f"Path already exists: {destination}")

        old_parent.remove_child(node.name)
        node.name = new_name
        node.parent = new_parent
        new_parent.add_child(node)


def main() -> None:
    fs = FileSystemManager()
    admin = User("root-admin", UserType.ADMIN)
    guest = User("guest", UserType.USER)

    fs.create_directory(admin, "/docs")
    fs.create_file(admin, "/docs/notes.txt")
    fs.write_file(admin, "/docs/notes.txt", "Meeting notes: LLD review at 3pm")

    fs.create_directory(admin, "/docs/drafts")
    fs.create_file(admin, "/docs/drafts/todo.txt")
    fs.write_file(admin, "/docs/drafts/todo.txt", "1. Finish parser\n2. Write tests")

    print(f"/docs size: {fs.get_size('/docs')} bytes")

    print("\nCreating an admin-only directory and writing to it as a guest:")
    fs.create_directory(admin, "/secure", required_role=UserType.ADMIN)
    fs.create_file(admin, "/secure/keys.txt")
    try:
        fs.write_file(guest, "/secure/keys.txt", "should not be allowed")
    except PermissionDeniedError as e:
        print(f"  {e}")

    print("\nWriting to it as an admin succeeds:")
    fs.write_file(admin, "/secure/keys.txt", "api-key-12345")
    print(f"  /secure/keys.txt -> '{fs.read_file(admin, '/secure/keys.txt')}'")

    print("\nMoving /docs/drafts/todo.txt to /docs/todo.txt:")
    fs.move(admin, "/docs/drafts/todo.txt", "/docs/todo.txt")
    print(f"  /docs/todo.txt -> '{fs.read_file(admin, '/docs/todo.txt')}'")

    print("\nDeleting /docs/drafts (now empty):")
    fs.delete(admin, "/docs/drafts")
    print(f"  /docs size after cleanup: {fs.get_size('/docs')} bytes")


if __name__ == "__main__":
    main()

File: Golang/Interview_problems/File_System_demo.go
/*
================================================================================
LLD: In-Memory File System
================================================================================

Functional Requirements:
    1. Create a directory at a given path.
    2. Create a file at a given path.
    3. Write (overwrite) content to a file.
    4. Read the content of a file.
    5. Delete a file or a directory (recursively).
    6. Calculate the size of a file or a directory (recursive sum).
    7. Move a file or directory from one path to another.
    8. Enforce permissions at the directory level for ADMIN vs USER.

Non-Functional Requirements:
    1. Extensibility: adding a new node type (e.g. symlink) shouldn't
       require touching path-resolution logic.
    2. Predictable errors: an operation on a path that doesn't exist, or
       one a user lacks permission for, fails with a clear exception
       instead of corrupting the tree.

Design:
    FileSystemNode is the common interface File and Directory both
    satisfy -- both know their own size. Directory additionally holds a
    name -> FileSystemNode map of children and a RequiredRole: an
    operation that mutates something under this directory (or the
    directory itself) must be performed by a user whose role is at least
    RequiredRole.

    FileSystemManager resolves "/"-separated paths by walking the tree
    from the root and performs permission checks against the *parent*
    directory of the node being mutated (create/write/delete/move all
    change the parent's contents).

Core Entities:
    UserType / User
    FileSystemNode (interface) / File / Directory
    FileSystemManager
================================================================================
*/

package main

import (
	"errors"
	"fmt"
	"strings"
)

type UserType int

const (
	RoleUser UserType = iota
	RoleAdmin
)

func (t UserType) String() string {
	if t == RoleAdmin {
		return "ADMIN"
	}
	return "USER"
}

type User struct {
	Name string
	Role UserType
}

type FileSystemNode interface {
	Name() string
	SetName(name string)
	Parent() *Directory
	SetParent(parent *Directory)
	GetSize() int
	IsDirectory() bool
}

type baseNode struct {
	name   string
	parent *Directory
}

func (n *baseNode) Name() string           { return n.name }
func (n *baseNode) SetName(name string)    { n.name = name }
func (n *baseNode) Parent() *Directory     { return n.parent }
func (n *baseNode) SetParent(p *Directory) { n.parent = p }

type File struct {
	baseNode
	Content string
}

func NewFile(name string, parent *Directory) *File {
	return &File{baseNode: baseNode{name: name, parent: parent}}
}

func (f *File) Read() string          { return f.Content }
func (f *File) Write(content string)  { f.Content = content }
func (f *File) Append(content string) { f.Content += content }
func (f *File) GetSize() int          { return len(f.Content) }
func (f *File) IsDirectory() bool     { return false }

type Directory struct {
	baseNode
	Children     map[string]FileSystemNode
	RequiredRole UserType
}

func NewDirectory(name string, parent *Directory, requiredRole UserType) *Directory {
	return &Directory{baseNode: baseNode{name: name, parent: parent},
		Children: make(map[string]FileSystemNode), RequiredRole: requiredRole}
}

func (d *Directory) AddChild(child FileSystemNode)       { d.Children[child.Name()] = child }
func (d *Directory) GetChild(name string) FileSystemNode { return d.Children[name] }
func (d *Directory) RemoveChild(name string)             { delete(d.Children, name) }

func (d *Directory) GetSize() int {
	total := 0
	for _, child := range d.Children {
		total += child.GetSize()
	}
	return total
}

func (d *Directory) IsDirectory() bool { return true }

var (
	ErrPathNotFound      = errors.New("path not found")
	ErrPathAlreadyExists = errors.New("path already exists")
	ErrPermissionDenied  = errors.New("permission denied")
)

func splitPath(path string) ([]string, error) {
	raw := strings.Split(path, "/")
	var parts []string
	for _, part := range raw {
		if part != "" {
			parts = append(parts, part)
		}
	}
	if len(parts) == 0 {
		return nil, errors.New("path must not be empty")
	}
	return parts, nil
}

type FileSystemManager struct {
	Root *Directory
}

func NewFileSystemManager() *FileSystemManager {
	return &FileSystemManager{Root: NewDirectory("/", nil, RoleUser)}
}

func (m *FileSystemManager) resolveParent(parts []string) (*Directory, error) {
	current := m.Root
	for _, part := range parts[:len(parts)-1] {
		child := current.GetChild(part)
		dir, ok := child.(*Directory)
		if child == nil || !ok {
			return nil, fmt.Errorf("%w: %s", ErrPathNotFound, part)
		}
		current = dir
	}
	return current, nil
}

func (m *FileSystemManager) resolve(path string) (FileSystemNode, error) {
	parts, err := splitPath(path)
	if err != nil {
		return nil, err
	}
	var current FileSystemNode = m.Root
	for _, part := range parts {
		dir, ok := current.(*Directory)
		if !ok {
			return nil, fmt.Errorf("%w: not a directory: %s", ErrPathNotFound, current.Name())
		}
		child := dir.GetChild(part)
		if child == nil {
			return nil, fmt.Errorf("%w: %s", ErrPathNotFound, path)
		}
		current = child
	}
	return current, nil
}

func (m *FileSystemManager) checkPermission(directory *Directory, user User) error {
	if user.Role < directory.RequiredRole {
		return fmt.Errorf("%w: user %s (%s) lacks permission for %s (requires %s)",
			ErrPermissionDenied, user.Name, user.Role, directory.Name(), directory.RequiredRole)
	}
	return nil
}

func (m *FileSystemManager) CreateDirectory(user User, path string, requiredRole UserType) (*Directory, error) {
	parts, err := splitPath(path)
	if err != nil {
		return nil, err
	}
	parent, err := m.resolveParent(parts)
	if err != nil {
		return nil, err
	}
	if err := m.checkPermission(parent, user); err != nil {
		return nil, err
	}
	name := parts[len(parts)-1]
	if parent.GetChild(name) != nil {
		return nil, fmt.Errorf("%w: %s", ErrPathAlreadyExists, path)
	}
	directory := NewDirectory(name, parent, requiredRole)
	parent.AddChild(directory)
	return directory, nil
}

func (m *FileSystemManager) CreateFile(user User, path string) (*File, error) {
	parts, err := splitPath(path)
	if err != nil {
		return nil, err
	}
	parent, err := m.resolveParent(parts)
	if err != nil {
		return nil, err
	}
	if err := m.checkPermission(parent, user); err != nil {
		return nil, err
	}
	name := parts[len(parts)-1]
	if parent.GetChild(name) != nil {
		return nil, fmt.Errorf("%w: %s", ErrPathAlreadyExists, path)
	}
	file := NewFile(name, parent)
	parent.AddChild(file)
	return file, nil
}

func (m *FileSystemManager) WriteFile(user User, path, content string) error {
	node, err := m.resolve(path)
	if err != nil {
		return err
	}
	file, ok := node.(*File)
	if !ok {
		return fmt.Errorf("not a file: %s", path)
	}
	if err := m.checkPermission(file.Parent(), user); err != nil {
		return err
	}
	file.Write(content)
	return nil
}

func (m *FileSystemManager) ReadFile(path string) (string, error) {
	node, err := m.resolve(path)
	if err != nil {
		return "", err
	}
	file, ok := node.(*File)
	if !ok {
		return "", fmt.Errorf("not a file: %s", path)
	}
	return file.Read(), nil
}

func (m *FileSystemManager) Delete(user User, path string) error {
	node, err := m.resolve(path)
	if err != nil {
		return err
	}
	parent := node.Parent()
	if parent == nil {
		return errors.New("cannot delete the root directory")
	}
	if err := m.checkPermission(parent, user); err != nil {
		return err
	}
	parent.RemoveChild(node.Name())
	return nil
}

func (m *FileSystemManager) GetSize(path string) (int, error) {
	node, err := m.resolve(path)
	if err != nil {
		return 0, err
	}
	return node.GetSize(), nil
}

func (m *FileSystemManager) Move(user User, source, destination string) error {
	node, err := m.resolve(source)
	if err != nil {
		return err
	}
	oldParent := node.Parent()
	if oldParent == nil {
		return errors.New("cannot move the root directory")
	}
	if err := m.checkPermission(oldParent, user); err != nil {
		return err
	}

	destParts, err := splitPath(destination)
	if err != nil {
		return err
	}
	newParent, err := m.resolveParent(destParts)
	if err != nil {
		return err
	}
	if err := m.checkPermission(newParent, user); err != nil {
		return err
	}
	newName := destParts[len(destParts)-1]
	if newParent.GetChild(newName) != nil {
		return fmt.Errorf("%w: %s", ErrPathAlreadyExists, destination)
	}

	oldParent.RemoveChild(node.Name())
	node.SetName(newName)
	node.SetParent(newParent)
	newParent.AddChild(node)
	return nil
}

func main() {
	fs := NewFileSystemManager()
	admin := User{Name: "root-admin", Role: RoleAdmin}
	guest := User{Name: "guest", Role: RoleUser}

	fs.CreateDirectory(admin, "/docs", RoleUser)
	fs.CreateFile(admin, "/docs/notes.txt")
	fs.WriteFile(admin, "/docs/notes.txt", "Meeting notes: LLD review at 3pm")

	fs.CreateDirectory(admin, "/docs/drafts", RoleUser)
	fs.CreateFile(admin, "/docs/drafts/todo.txt")
	fs.WriteFile(admin, "/docs/drafts/todo.txt", "1. Finish parser\n2. Write tests")

	size, _ := fs.GetSize("/docs")
	fmt.Printf("/docs size: %d bytes\n", size)

	fmt.Println("\nCreating an admin-only directory and writing to it as a guest:")
	fs.CreateDirectory(admin, "/secure", RoleAdmin)
	fs.CreateFile(admin, "/secure/keys.txt")
	if err := fs.WriteFile(guest, "/secure/keys.txt", "should not be allowed"); err != nil {
		fmt.Printf("  %s\n", err)
	}

	fmt.Println("\nWriting to it as an admin succeeds:")
	fs.WriteFile(admin, "/secure/keys.txt", "api-key-12345")
	content, _ := fs.ReadFile("/secure/keys.txt")
	fmt.Printf("  /secure/keys.txt -> '%s'\n", content)

	fmt.Println("\nMoving /docs/drafts/todo.txt to /docs/todo.txt:")
	fs.Move(admin, "/docs/drafts/todo.txt", "/docs/todo.txt")
	content, _ = fs.ReadFile("/docs/todo.txt")
	fmt.Printf("  /docs/todo.txt -> '%s'\n", content)

	fmt.Println("\nDeleting /docs/drafts (now empty):")
	fs.Delete(admin, "/docs/drafts")
	size, _ = fs.GetSize("/docs")
	fmt.Printf("  /docs size after cleanup: %d bytes\n", size)
}
